### PR TITLE
Adding support for Pop!_OS

### PR DESCRIPTION
**Pull request by Emprario**
Adding Pop!_OS support in eupnea (building from iso). Indeed, to build Pop!_OS, the script downloads the iso from the Pop!_OS website then extracts the squashfs from the iso to /mnt/eupnea. In a last time, the script removes packages and configures the installation like the Pop!_OS installer.

### CODE REVIEW

File: build.py [hunk 1]
```diff
@@ -150,6 +173,15 @@ def download_rootfs(distro_name: str, distro_version: str, distro_link: str) ->
                 start_download_progress("/tmp/eupnea-build/fedora-rootfs.raw.xz")
                 urlretrieve(distro_link, filename="/tmp/eupnea-build/fedora-rootfs.raw.xz")
                 stop_download_progress()
+            case "pop-os":
+                print_status(f"Downloading Pop!_OS iso {distro_version}")
+                if not path_exists("/tmp/persistant/pop-os.iso"):
```
Comment (Apacelus) [inline build.py:178]: The script should always download fresh images, unless --local-path is used

File: build.py [hunk 2]
```diff
@@ -432,6 +475,9 @@ def post_extract(build_options, kernel_type: str) -> None:
                 chroot(f"usermod -aG sudo {username}")
             case "arch" | "fedora":
                 chroot(f"usermod -aG wheel {username}")
+            case "pop-os":
+                sleep(5) # need to sleep for some reasons
```
Comment (Apacelus) [inline build.py:479]: are you sure thats absolutely necessary?

File: build.py [hunk 3]
```diff
@@ -432,6 +475,9 @@ def post_extract(build_options, kernel_type: str) -> None:
...
+                chroot(f"usermod -aG adm,sudo,lpadmin {username}")
```
Comment (Apacelus) [inline build.py:480]: Does the popos installer add those groups? Because stock Ubuntu does not need them...

File: build.py [hunk 4]
```diff
@@ -291,6 +323,7 @@ def partition_and_flash_kernel(mnt_point: str, write_usb: bool, distro_name: str
 
 # extract the rootfs to /mnt/eupnea
 def extract_rootfs(distro_name: str) -> None:
+    global mnt_iso # pop-os support
```
Comment (Apacelus) [inline build.py:326]: Rename to popos_iso_mnt for better code clarity

File: cli_input.py [hunk 5]
```diff
@@ -85,71 +85,79 @@ def get_user_input() -> dict:
                             print_warning("Version not available, please choose another")
                             continue
                 break
+            case "Pop!_OS"|"PopOS"|"POP!_OS"|"Pop_OS"|"Pop!OS"|"pop!_os":
```
Comment (Apacelus) [inline cli_input.py:88]: add "pop-os"

File: build.py (hunk 3 above)
Comment (Apacelus) [reply build.py:480]: Hmm, I suppose. Does Ubuntu perhaps need them too?

Comment (WeirdTreeThing) [reply build.py:480]: You should add them to every distro to make sure the user has access to all the hardware they need.

Comment (WeirdTreeThing) [reply build.py:480]: Include ones like video audio and input.

Comment (Apacelus) [reply build.py:480]: Hmm, but even on a stock popos install, a user is not in as many groups. I dont think its a good idea to add the user to so many groups for "good measure" if a mainline install doesnt do that either

File: distro/popos.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+def config(de_name: str, distro_version: str, root_partuuid: str, verbose: bool) -> None:
+    set_verbose(verbose)
+    print_status("Configuring Pop!_OS")
+
+    print_status("Removing casper debs")
+    chroot("apt-get purge btrfs-progs casper cifs-utils distinst distinst-v2 dmraid expect f2fs-tools fatresize gettext gparted gparted-common grub-common grub2-common kpartx kpartx-boot libdistinst libdmraid1.0.0.rc16 libinih1 libnss-mymachines localechooser-data os-prober pop-installer pop-installer-casper pop-shop-casper squashfs-tools systemd-container tcl-expect user-setup xfsprogs -y")
+
+    # Replace input-synaptics with newer input-libinput, for better touchpad support
+    #print_status("Upgrading touchpad drivers")
+    #chroot("apt-get remove -y xserver-xorg-input-synaptics")
```
Comment (Apacelus) [inline distro/popos.py:13]: Why is this commented? Does it break something in PopOS?

File: build.py (hunk 4 above)
Comment (Apacelus) [inline build.py:326]: No need for global var, just unmount immediately after use (see next comment)

File: build.py [hunk 7]
```diff
@@ -546,6 +596,9 @@ def start_build(verbose: bool, local_path: str, kernel_type: str, dev_release: b
     print_status("Cleaning up host system after build")
     try:
         bash("umount -f /mnt/eupnea")
+        if build_options["distro_name"]=="pop-os":
+            bash("umount -f /tmp/eupnea-build/cdrom") # pop-os loop device
+            bash(f"losetup -d {mnt_iso} ")
     except subprocess.CalledProcessError:  # on crostini umount fails for some reason
         pass
```
Comment (Apacelus) [inline build.py:603]: No need for this, as image is unmounted immediately after use

File: build.py [hunk 8]
```diff
@@ -546,6 +596,9 @@ def start_build(verbose: bool, local_path: str, kernel_type: str, dev_release: b
     print_status("Cleaning up host system after build")
     try:
         bash("umount -f /mnt/eupnea")
+        if build_options["distro_name"]=="pop-os":
+            bash("umount -f /tmp/eupnea-build/cdrom") # pop-os loop device
+            bash(f"losetup -d {mnt_iso} ")
```
Comment (Apacelus) [inline build.py:601]: No need for this, as image is unmounted immediately after use

File: build.py [hunk 9]
```diff
@@ -335,6 +368,16 @@ def extract_rootfs(distro_name: str) -> None:
...
             bash(f"losetup -d {fedora_root_part[:-2]}")
+        case "pop-os":
+            print_status("Extracting Pop!_OS squashfs from iso")
+            # Create a mount point for the iso to extract the squashfs
+            mkdir("/tmp/eupnea-build/iso")
+            mnt_iso=bash(f"losetup -f --show /tmp/persistant/pop-os.iso")
+            mkdir("/tmp/eupnea-build/cdrom")
+            bash(f"mount {mnt_iso} /tmp/eupnea-build/cdrom")
+            bash("unsquashfs -f -d /mnt/eupnea /tmp/eupnea-build/cdrom/casper/filesystem.squashfs")
+
```
Comment (Apacelus) [inline build.py:379]: Just umount the image right here with:
```
# unmount pop-os loop device
bash("umount -f /tmp/eupnea-build/cdrom")
bash(f"losetup -d {mnt_iso} ")
```

File: distro/popos.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+def config(de_name: str, distro_version: str, root_partuuid: str, verbose: bool) -> None:
+    set_verbose(verbose)
+    print_status("Configuring Pop!_OS")
+
+    print_status("Removing casper debs")
+    chroot("apt-get purge btrfs-progs casper cifs-utils distinst distinst-v2 dmraid expect f2fs-tools fatresize gettext"
+           " gparted gparted-common grub-common grub2-common kpartx kpartx-boot libdistinst libdmraid1.0.0.rc16 "
+           "libinih1 libnss-mymachines localechooser-data os-prober pop-installer pop-installer-casper pop-shop-casper "
+           "squashfs-tools systemd-container tcl-expect user-setup xfsprogs -y")
```
Comment (Apacelus) [inline distro/popos.py:12]: Is this taken from a popos installer script?(Please add a comment into the script with the source if so)  

Coz its removing quite useful packages like kpartx, gettext, fatresize, systemd-container...

File: distro/popos.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
```
Comment (Apacelus) [inline distro/popos.py:13]: Please add cloud-utils install, as its needed for automatic resizing on first boot

File: distro/popos.py (hunk 10 above)
Comment (Apacelus) [reply distro/popos.py:12]: @Emprario can you please find the actual source code for this?

File: build.py (hunk 2 above)
Comment (Apacelus) [reply build.py:481]: @Emprario

File: build.py (hunk 1 above)
Comment (Emprario) [reply build.py]: Fixed.

File: build.py (hunk 2 above)
Comment (Emprario) [reply build.py:481]: Yes, it's necessary, without it didn't work for me (usermod wasn't find).

File: build.py (hunk 3 above)
Comment (Emprario) [reply build.py:482]: That's what done the installer from stock iso I think we should keep them.

File: cli_input.py [hunk 12]
```diff
@@ -30,7 +30,7 @@ def get_user_input() -> dict:
     print_question("Which Linux distro(flavor) would you like to use?")
 
     while True:
-        temp_distro_name = input("\033[94m" + "Available options: Ubuntu(default, recommended), Debian, Arch, Fedora\n"
+        temp_distro_name = input("\033[94m" + "Available options: Ubuntu(default, recommended), Debian, Arch, Fedora, Pop!_OS (still WIP)\n"
```
[the remark this answers is not in the document]
Comment (Emprario) [reply cli_input.py]: Not it's just an omission.

File: distro/popos.py (hunk 6 above)
Comment (Emprario) [reply distro/popos.py]: I uncomment and it works with my chromebook. I wasn't add it in a first place beceause libinput is something for wayland and pop-os is still on Xorg.

File: build.py (hunk 4 above)
Comment (Emprario) [reply build.py]: Fixed.

File: build.py (hunk 9 above)
Comment (Emprario) [reply build.py:381]: Done.

File: distro/popos.py (hunk 10 above)
Comment (Emprario) [reply distro/popos.py:12]: Yes it's taken from the installer, I write it down just above (in the script).

Comment (Emprario) [reply distro/popos.py:12]: @Apacelus It comes from the installer. So it's surely comes from somewhere here https://github.com/pop-os/distinst
Should I note it on the popos.py ?

Comment (Emprario) [reply distro/popos.py:12]: The problem is that I absolutly don't known where this code comes from extactly I just see it in the logs on my chromebook (the log that was provided the installer). Even by building myself this tool and install on Pop!_OS works but I never see any file (excepts the logs) where I can found this information.

Comment (Apacelus) [reply distro/popos.py:12]: Hmm, alright, just link the whole repo then.

Comment (Emprario) [reply distro/popos.py:12]: Done.

**Comment by Apacelus** [comment]
Alright, all looks good. Thank you for the pr!

**Pull request merged**